### PR TITLE
Working demo

### PR DESCRIPTION
**Pull request by arkgil**
This PR introduces enough bits and pieces to run a full monitoring scenario, i.e. create a metric and its view, attach a reporter and send metric somewhere (console in this case).

The simplest demo would be:

```
iex -S mix
iex> Metrics.add_reporter(Metrics.Reporters.Console)
iex> Metrics.Probes.Memory.start
```

You'll see some measurements of memory usage reported in the console every ten seconds. Now on to the hard stuff..

Right below you'll find a glossary section. Later on there is a "How it works" describing some architectural choices. Note that this is a proof of concept, everything is a subject to change.

## Glossary

Drawing from the tradition of all Erlang libraries, we need to have at least 3 pages long glossary at top of the README. This is a humble attempt to achieve this 😄 

### Metric

Metric is a collection of measurements. That's it. Is has a name, a description, maybe a unit, but it's an abstract collection of values. 

Measurement is always a number. How the collection of measurements (i.e. metric) is represented depends on *views* subscribed to it.

We should allow easy exploration of metrics, especially given that they will have descriptions.

### View

View receives measurements of a single metric and aggregates them. Aggregation may have many forms, e.g. a view might emit only the last recorded value (gauge), the rate of measurements recorded per unit of time, or quantiles, mean, min and max (the so called "histogram", although I don't really like this name). Each view performs only a single type of aggregation, but the result of aggregation might consist of many values (like in the example of histogram).

View chooses when to emit aggregated values which are **pushed** to *reporters*. This is different than most metric systems, usually values are pulled from aggregators. However we figured that in most cases it doesn't make sense to have independent metric interval (if you have a good example when that's needed, please tell!). Obviously this comes with the possibility of flooding reporters with  values - but we'll know when we measure/profile/benchmark. Recordings of metrics are the source of load in the system anyway, not the views emitting values.

View may be subscribed only to a single metric, but a single metric may have many views subscribed to it. Views also have a name, description, and maybe a unit.

Similarily to metrics, views also should be easily explorable. We should also define a behaviour, so that it's easy to create own views.

I'm not sure if views should be reconfigurable at runtime. Does it makes sense to change the size of the sliding window over which percentiles are calculated?

### Reporters

Reporters are quite self-explanatory, they report aggregated values to external systems. Reporters should have some kind of indentifier, so that we could reconfigure them in runtime (similar to Logger backends). Reporters should be able to subscribe to all metrics, selected metrics, or selected views.

There should be (and already is! a simple one, though) a behaviour for reporters. 

### Probes

Probe is a thing which periodically makes and records a measurement. In this PR there is a memory usage probe. There is nothing fancy going on with them, I'm not sure if there should be some kind of behaviour for them or just a name.

# How it works

Given that we defined metric as an abstract concept, the only thing we need to know about them is that if anyone's interested in their measurements (there is currently no place in code where the metric is registered etc.). Because of that, I tried to keep it simple and in this version you can only define views. 

Created views are kept in an ETS table. I've used the solution implemented in exometer, where there is an ETS table per scheduler, and the process always uses the table assigned to the scheduler it's currently running on (`Metrics.FastTables`). This means that we need to put a view registration info in each of those tables. To preserve "atomicity" view registration always goes through one process. This also allows us to reject registration of views with duplicate names. This process (`Metrics.ViewStore`) also monitors the process which performs a registration and removes views defined by it when it goes down. Currently we have only gauge views, and there is a single, dedicated process which does registration. I imagine that in the future when we have stateful views, like "histogram" or "rate" they would register themselves in the store.

When someone records a measurement, we look up views subscribed to it (in a table assigned to scheduler the process is running on) and call a `handle_measurement` callback (there isn't really a behaviour for that yet, because I imagine that we stateful views we would need more callbacks anyway). In case of gauges, recorded measurements are immediately pushed to reporters subscribed to that metric (more on reporter subcriptions below). We chose the route with pushing values to reporters, because it simply makes sense that the aggregator chooses when to emit a value. In case of gauges it doesn't make sense to report values every minute - why would I use gauge if I lose most of the values this way? The same goes for views which calculate some stats over the sliding window - it doesn't make much sense to report every 15 seconds, if I aggregate values from the last minute - it even looks confusing on the plotted chart. Again, if anyone know a good use case for having separate report interval, I'd be happy to hear it 🙂 
The next view I wanted to implemented was a counter. Basically we would have a per-counter entry in all ETS tables, and have a dedicated process which periodically sums the counters, resets them and pushes to reporters. This process would also perform a registration in a `ViewStore`.

The other thing about views - I think that we could document it somewhere that it rarely makes sense to attach multiple views of different types to the same metric. The only example I was thinking of was when measurement is a request latency and we can have one view calculating quantiles, and the other view counting the rate of requests. But I'd still say that these should be some separate metrics. Multiple views of the same metric might be useful, for example, when we want to have a counter of events over last minute, but also since the start of the system. That would be a single metric, with two views of the same type, but with different configurations.

Reporters are also registered in these "fast" ETS tables. However, currently there is no way to control which views or metrics they are subscribed to.

### CODE REVIEW

File: lib/metrics/reporter.ex [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  use GenServer
+
+  alias Metrics.{ReporterSupervisor, ReporterStore}
+
+  @doc """
+  Called whenever any of the views emits its values
+
+  The last argument is the UTC timestamp when the values have been emitted.
+  """
+  @callback handle_emit(Metrics.metric(), Metrics.view(), [Metrics.value()], NaiveDateTime.t()) ::
```
Comment (josevalim) [inline lib/metrics/reporter.ex]: I would keep the last argument the unix timestamp for simplicity. I even think Erlang's `logger` in OTP 21 did a similar change?

File: lib/metrics/view_store.ex [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  @spec remove_fast_tables_entries([Metrics.view()]) :: any
+  defp remove_fast_tables_entries(views), do: Enum.each(views, &remove_fast_tables_entry/1)
+
+  @spec remove_fast_tables_entry(Metrics.view()) :: any
+  defp remove_fast_tables_entry(view) do
+    pattern = {{:view, :_}, view, :_, :_}
+
+    for table <- FastTables.get_all() do
+      [entry] = :ets.match_object(table, pattern)
+      :ets.delete_object(table, entry)
```
Comment (josevalim) [inline lib/metrics/view_store.ex:102]: You can use match_delete here I think?

**Comment by josevalim** [comment]
Thanks @arkgil!

I have done an initial review and I believe my only concern right now is that the Gauge is a centralized process. I am also not sure if we should force reporters to be a separate process either.

I have been thinking about this and I am starting to get the conclusion that views are actually a property of the reporter and not a general mechanism. For example, if we think about the controller/view instrumentation that we have in Phoenix, I want to hook into each of those events directly, without going through a view+reporter. At the same time, I still want to be able to define a set of views to control how data is pushed to prometheus or to a sample page in Phoenix.

So here is what I propose (and this is pretty much open to discussion):

  1. Let's focus on a measurement system (collector) with publishers. The design here can be really similar to how logger in OTP 21 works: each publisher is registered in an ETS table and they are invoked inline and directly where the measurement happens. If a publisher fails, it is removed from the table.

  2. We will provide a reporter which is a built-in publisher that works with views. The goal of the views is to have a high-level language to describe how data is collected, then the reporter is capable of pushing those views somewhere else. We can keep the aggregation of each view per reporter and simplify the design.

Thoughts?

**Comment by josevalim** [comment]
Ping. :)

**Comment by rslota** [comment]
Hi @josevalim ,

I just wanted to share my 2 cents :)

> Let's focus on a measurement system (collector) with publishers. The design here can be really similar to how logger in OTP 21 works: each publisher is registered in an ETS table and they are invoked inline and directly where the measurement happens. If a publisher fails, it is removed from the table.

I'm sceptical about this a little. It would create back-off based on the slowest reporter. In my opinion would be better to isolate reporters' event sinks in order to allow faster reporters to get more events even if there is other reporter configured, that is not able to process so many measurements. e.g., `InfluxDB` TCP based reporter is significantly slower then simplest `statsd` UDP based reporter and when both configured, we should make sure, that the faster one will get all events it can handle even if the slower one struggles.

In short - I think that back-off should be on per reporter basis.

**Comment by arkgil** [comment]
Sorry @josevalim , I somehow missed the notification in the load of other emails.

Gauge here is not a centralized process (the code does not really have top quality, so you might have missed it) - centralized process only performs registration of the gauge view, but measurements are pushed directly by the recording process to subscribed reporters. 

I believe that the instrumentation that Phoenix provides should be left in Phoenix and kept completely outside the scope of this project. What I was thinking about is that Phoenix will install its own views and record metrics, and I, as a user, will be able to hook my reporter into those views. So from the user of the library POV it's very easy to hook up into metrics exposed by it.

As I understand it, metric is just a stream of measurements, numbers. VIews (but now I think a better name would be aggregator or materializer) feeds off this stream and materializes it as concrete values. The gauge aggregator materializes each measurement as a separate value, while "histogram" aggregator materializes the stream as statistic (min, max, quantiles) about measurements from, say, last minute. Reporters in this case are only responsible for pushing materialized value out - be it Prometheus, Phoenix metrics page, console etc.

That's why I think that the "view" is not a property of a reporter. "View" defines what data we actually see - since metric is just an infinite stream of measurements, we need to materialize it somehow. Reporter on the other hand defines only where we can see those generated values, and that's why I think it's completely separate from the view. I could have a view which emits quantiles of request latency, but I'd like to see that data both in console and in Graphite - thus two distinct reporters. With this design views work exactly as how you described publishers - they are stored in ETS table and each is called sequentially when the measurement is recorded. Going back to example of gauge, when invoked it simply emits the value to reporters directly, without going through any other process, but I imagine that stateful aggregations like "histogram" would require another process.

WIth this kind of design reporters need to be in a separate process, because when the values are aggregated and emitted is completely irrelevant of when the measurement is recorded.

To sum up:

1. We already have a collection and "publishers" mechanism in place. That's the `Metrics.record/2` function and `ViewStore` module  used for registering and looking up views (publishers). 
2. We should provide a set of views which are most common - gauges, counters, histograms, etc. With mechanism from point 1. anyone can build their own custom view and hook it up to handle measurements in non-standard way if needed.
3. We should provide utilities for building reporters, as well as for starting them and subscribing to selected views. There should also be a simple console reporter built in.

**Comment by josevalim** [comment]
Hi @rslota! What you have described is definitely possible to happen. :) That's why the recommendation should be for each reporter to do as minimal sync work as possible. That's the recommendation for Erlang/OTP 21 loggers too.

My concern with us prescribing how events are published is that all of the rules regarding back-pressure will be up for us to implement and they may be unable to fit all possible use cases.

**Comment by arkgil** [comment]
@josevalim after some more thought I agree that reporters could be called synchronously just like views are. Of course if that function sends a message, that's not a problem. Is that what you meant?

**Comment by josevalim** [comment]
> Gauge here is not a centralized process (the code does not really have top quality, so you might have missed it) - centralized process only performs registration of the gauge view, but measurements are pushed directly by the recording process to subscribed reporters.

Apologies for missing this. I will take another look at the code then.

> I believe that the instrumentation that Phoenix provides should be left in Phoenix and kept completely outside the scope of this project.

I would like for us to reconsider this. The appeal of this project is to provide a unified toolset on how data is gathered. If we ship this project, and then Phoenix still has its own instrumentation, and Ecto still has its own instrumentation, and so on, then we will fail at our goal of having one and only one mechanism of tracking instrumentations and metrics in the system.

> That's why I think that the "view" is not a property of a reporter. "View" defines what data we actually see - since metric is just an infinite stream of measurements, we need to materialize it somehow. 

To be clear, I completely agree with this. My points here are:

  * Different reporters may want to show different views, different materialization of the data
  * If you have a view but nobody reporter to consume it, does this view even matter?

So views should be their own concept but, in my mind, without a reporter, a view is not very useful. So how to best model this relationship?

However, I believe we should hold the view discussion for now since the matter of considering Phoenix instrumentation or not is much more central to the design of the library and whatever we decide there will affect this part of the system.

**Comment by josevalim** [comment]
If you folks want to jump on call, just let me know. :)

**Comment by arkgil** [comment]
@josevalim sure, Skype? I'm `arkadiusz.gil.esl`

**Comment by arkgil** [comment]
To sum up what has been discussed on the call:
1. We're gonna ship a core events systems, where anyone can emit events and subscribe to all or a subset of them. This will replace Phoenix built-in instrumentation, and instead relevant events will be published to that system. Event's values will most likely be numerical, but might be also something else if anyone wants to do that.
2. On top of that we're going to build a system for aggregating and reporting metrics. How exactly we're going to build it (e.g. if reporters are separate from views etc.) is still an open discussion.

Currently we're gonna focus on point 1.

**Comment by josevalim** [comment]
Btw, [Statix](https://github.com/lexmag/statix) is a nice example of why to use ETS. They don't have processes and they write to the socket directly, which performs way better than going through a process. For [InfluxDB metrics, they use a pool](https://github.com/lexmag/fluxter).

Another suggestion from @lexmag is to use a list of atoms for the event names. The issue with using `"foo.bar.baz"` is that different tools use different separators. Going with a list keeps it more generic and allows for faster matching.

**Comment by arkgil** [comment]
@josevalim yes, list of atoms might not look as pretty in code as string with dot-separated segments, but it's a more pragmatic choice. That's what exometer and folsom do as well.

And yes, ETS will probably be the most efficient method. What I'd like the measure though is the tradeoff of memory vs performance of two implementations: single table with read concurrency and table per scheduler.

**Comment by cloud8421** [comment]
Just caught up on the entire thread, just wanna clarify one point.

@arkgil said:

> We're gonna ship a core events systems, where anyone can emit events and subscribe to all or a subset of them. This will replace Phoenix built-in instrumentation, and instead relevant events will be published to that system. Event's values will most likely be numerical, but might be also something else if anyone wants to do that.

This assumes what @josevalim said earlier:

> Let's focus on a measurement system (collector) with publishers. The design here can be really similar to how logger in OTP 21 works: each publisher is registered in an ETS table and they are invoked inline and directly where the measurement happens. If a publisher fails, it is removed from the table.

Is that right?

I'm personally in favor of this inlined approach as it makes it easy to measure the impact of metrics collection on the system.

In addition, one of the problems that I regularly face in similar situations (gathering metrics, logging, etc.) is uncontrollable pressure caused by process communication that I can't control because it's buried in library code I can't modify unless I fork the project itself.

**Comment by arkgil** [comment]
Yes @cloud8421 , that's correct 🙂
